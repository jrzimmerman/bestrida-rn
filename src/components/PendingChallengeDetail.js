import React from 'react';
import {
  Dimensions,
  View,
  StatusBar,
  Text,
  TouchableOpacity
} from 'react-native';
import PropTypes from 'prop-types';
import { connect } from 'react-redux';
import SegmentMap from './SegmentMap';
import styles from '../styles/styles';
import pendingStyles from '../styles/pendingStyles';
import * as challengeActions from '../actions/challenges';
import * as segmentActions from '../actions/segments';

export class PendingChallengeDetail extends React.Component {
  constructor(props) {
    super(props);
    this.handleAccept = this.handleAccept.bind(this);
    this.handleDecline = this.handleDecline.bind(this);
  }

  componentDidMount() {
    const { dispatch } = this.props;
    const { challenge } = this.props.navigation.state.params;
    dispatch(segmentActions.getSegment(challenge.segment.id));
  }

  componentWillUnmount() {
    this.props.dispatch(segmentActions.clearSegment());
  }

  handleAccept(challengeId, userId) {
    const { dispatch, navigation } = this.props;
    dispatch(challengeActions.acceptChallenge(challengeId, userId));
    dispatch(challengeActions.pendingChallenges(userId));
    navigation.goBack();
  }

  handleDecline(challengeId, userId) {
    const { dispatch, navigation } = this.props;
    dispatch(challengeActions.declineChallenge(challengeId, userId));
    dispatch(challengeActions.pendingChallenges(userId));
    navigation.goBack();
  }

  render() {
    const { segments } = this.props;
    const { challenge, userId } = this.props.navigation.state.params;
    const { height, width } = Dimensions.get('window');
    let segmentMap;
    if (segments && segments.segment && segments.segment.map) {
      segmentMap = (
        <View style={styles.challengeMapView}>
          <SegmentMap
            height={height * 0.275}
            width={width * 0.9}
            map={segments.segment.map}
          />
        </View>
      );
    }

    let challengeFooter;
    if (userId && challenge.challengee.id === userId) {
      challengeFooter = (
        <View style={pendingStyles.challengeOptions}>
          <TouchableOpacity
            onPress={() => this.handleDecline(challenge.id, userId)}
            style={pendingStyles.challengeOptionsDecline}
          >
            <Text style={pendingStyles.challengeOptionsDeclineText}>
              Decline
            </Text>
          </TouchableOpacity>
          <TouchableOpacity
            onPress={() => this.handleAccept(challenge.id, userId)}
            style={pendingStyles.challengeOptionsAccept}
          >
            <Text style={pendingStyles.challengeOptionsAcceptText}>Accept</Text>
          </TouchableOpacity>
        </View>
      );
    } else if (userId && challenge.challenger.id === userId) {
      challengeFooter = (
        <TouchableOpacity
          onPress={() => this.handleDecline(challenge.id, userId)}
          style={styles.button}
        >
          <Text style={styles.buttonText}>Cancel Challenge</Text>
        </TouchableOpacity>
      );
    }

    return (
      <View style={styles.container}>
        <StatusBar barStyle="light-content" />
        <View style={styles.challengeTitleView}>
          <Text style={styles.challengeTitleText}>
            {challenge.segment.name}
          </Text>
        </View>
        <View style={styles.challengeDetailView}>
          <View style={styles.detailRowView}>
            <Text style={styles.challengeDetailTitle}>Start Date</Text>
            <Text style={styles.challengeDetailText}>
              {new Date(challenge.created).toDateString()}
            </Text>
          </View>
          <View style={styles.detailRowView}>
            <Text style={styles.challengeDetailTitle}>End Date</Text>
            <Text style={styles.challengeDetailText}>
              {new Date(challenge.expires).toDateString()}
            </Text>
          </View>
          <View style={styles.detailRowView}>
            <Text style={styles.challengeDetailTitle}>Distance</Text>
            <Text style={styles.challengeDetailText}>
              {`${(challenge.segment.distance / 1609.34).toFixed(2)} Miles` ||
                '0'}
            </Text>
          </View>
          <View style={styles.detailRowView}>
            <Text style={styles.challengeDetailTitle}>Location</Text>
            <Text style={styles.challengeDetailText}>
              {`${challenge.segment.city
                ? `${challenge.segment.city},`
                : ''} ${challenge.segment.state}`}
            </Text>
          </View>
          <View style={styles.detailRowView}>
            <Text style={styles.challengeDetailTitle}>Activity Type</Text>
            <Text style={styles.challengeDetailText}>
              {challenge.segment.activityType || 'Not Available'}
            </Text>
          </View>
          <View style={styles.detailRowView}>
            <Text style={styles.challengeDetailTitle}>Average Grade</Text>
            <Text style={styles.challengeDetailText}>{`${challenge.segment
              .averageGrade} %`}</Text>
          </View>
          <View style={styles.detailRowView}>
            <Text style={styles.challengeDetailTitle}>Climb Category</Text>
            <Text style={styles.challengeDetailText}>
              {challenge.segment.climbCategory || 'Not Available'}
            </Text>
          </View>
          <View style={styles.detailRowView}>
            <Text style={styles.challengeDetailTitle}>Elevation Gain</Text>
            <Text style={styles.challengeDetailText}>
              {`${challenge.segment.totalElevationGain} meters`}
            </Text>
          </View>
        </View>
        {segmentMap}
        <View style={styles.challengeFooterView}>{challengeFooter}</View>
      </View>
    );
  }
}

const { func, number, shape } = PropTypes;

PendingChallengeDetail.propTypes = {
  dispatch: func,
  challenge: shape({}),
  userId: number,
  navigation: shape({}),
  segments: shape({})
};

const mapStateToProps = state => ({
  userId: state.user.auth.userId,
  segments: state.segments
});

export default connect(mapStateToProps)(PendingChallengeDetail);
