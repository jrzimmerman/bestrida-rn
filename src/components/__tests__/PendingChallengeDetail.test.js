import 'react-native';
import React from 'react';
import { PendingChallengeDetail } from '../PendingChallengeDetail';
import configureMockStore from 'redux-mock-store';
import thunk from 'redux-thunk';
// Note: test renderer must be required after react-native.
import renderer from 'react-test-renderer';

const middlewares = [thunk];
const mockStore = configureMockStore(middlewares);

test('renders PendingChallengeDetail component', () => {
  const challenge = {
    segment: {
      name: 'Test Segment'
    }
  };
  const navigation = {
    state: {
      params: {
        challenge
      }
    }
  };
  const store = mockStore({});
  const tree = renderer
    .create(
      <PendingChallengeDetail
        navigation={navigation}
        dispatch={fn => fn}
        store={store}
      />
    )
    .toJSON();
  expect(tree).toMatchSnapshot();
});
