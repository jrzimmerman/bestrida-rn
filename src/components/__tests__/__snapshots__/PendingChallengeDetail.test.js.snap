// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders PendingChallengeDetail component 1`] = `
<View
  style={
    Object {
      "alignItems": "center",
      "backgroundColor": "#2B2B2B",
      "flex": 1,
    }
  }
>
  <View
    style={
      Object {
        "alignSelf": "stretch",
        "flex": 0.1,
      }
    }
  >
    <Text
      accessible={true}
      allowFontScaling={true}
      disabled={false}
      ellipsizeMode="tail"
      style={
        Object {
          "alignSelf": "center",
          "color": "#CCC",
          "flexDirection": "row",
          "fontSize": 18,
          "fontWeight": "bold",
          "justifyContent": "space-around",
          "paddingTop": 20,
        }
      }
    >
      Test Segment
    </Text>
  </View>
  <View
    style={
      Object {
        "alignSelf": "stretch",
        "backgroundColor": "#383838",
        "flex": 0.375,
        "marginBottom": 10,
        "marginHorizontal": 20,
      }
    }
  >
    <View
      style={
        Object {
          "alignSelf": "stretch",
          "flex": 1,
          "flexDirection": "row",
          "justifyContent": "space-between",
          "marginHorizontal": 20,
          "marginTop": 5,
        }
      }
    >
      <Text
        accessible={true}
        allowFontScaling={true}
        disabled={false}
        ellipsizeMode="tail"
        style={
          Object {
            "alignSelf": "center",
            "color": "#CCC",
            "flexDirection": "column",
            "fontSize": 14,
            "fontWeight": "bold",
          }
        }
      >
        Start Date
      </Text>
      <Text
        accessible={true}
        allowFontScaling={true}
        disabled={false}
        ellipsizeMode="tail"
        style={
          Object {
            "alignSelf": "center",
            "color": "#CCC",
            "flexDirection": "column",
            "fontSize": 12,
          }
        }
      >
        Invalid Date
      </Text>
    </View>
    <View
      style={
        Object {
          "alignSelf": "stretch",
          "flex": 1,
          "flexDirection": "row",
          "justifyContent": "space-between",
          "marginHorizontal": 20,
          "marginTop": 5,
        }
      }
    >
      <Text
        accessible={true}
        allowFontScaling={true}
        disabled={false}
        ellipsizeMode="tail"
        style={
          Object {
            "alignSelf": "center",
            "color": "#CCC",
            "flexDirection": "column",
            "fontSize": 14,
            "fontWeight": "bold",
          }
        }
      >
        End Date
      </Text>
      <Text
        accessible={true}
        allowFontScaling={true}
        disabled={false}
        ellipsizeMode="tail"
        style={
          Object {
            "alignSelf": "center",
            "color": "#CCC",
            "flexDirection": "column",
            "fontSize": 12,
          }
        }
      >
        Invalid Date
      </Text>
    </View>
    <View
      style={
        Object {
          "alignSelf": "stretch",
          "flex": 1,
          "flexDirection": "row",
          "justifyContent": "space-between",
          "marginHorizontal": 20,
          "marginTop": 5,
        }
      }
    >
      <Text
        accessible={true}
        allowFontScaling={true}
        disabled={false}
        ellipsizeMode="tail"
        style={
          Object {
            "alignSelf": "center",
            "color": "#CCC",
            "flexDirection": "column",
            "fontSize": 14,
            "fontWeight": "bold",
          }
        }
      >
        Distance
      </Text>
      <Text
        accessible={true}
        allowFontScaling={true}
        disabled={false}
        ellipsizeMode="tail"
        style={
          Object {
            "alignSelf": "center",
            "color": "#CCC",
            "flexDirection": "column",
            "fontSize": 12,
          }
        }
      >
        NaN Miles
      </Text>
    </View>
    <View
      style={
        Object {
          "alignSelf": "stretch",
          "flex": 1,
          "flexDirection": "row",
          "justifyContent": "space-between",
          "marginHorizontal": 20,
          "marginTop": 5,
        }
      }
    >
      <Text
        accessible={true}
        allowFontScaling={true}
        disabled={false}
        ellipsizeMode="tail"
        style={
          Object {
            "alignSelf": "center",
            "color": "#CCC",
            "flexDirection": "column",
            "fontSize": 14,
            "fontWeight": "bold",
          }
        }
      >
        Location
      </Text>
      <Text
        accessible={true}
        allowFontScaling={true}
        disabled={false}
        ellipsizeMode="tail"
        style={
          Object {
            "alignSelf": "center",
            "color": "#CCC",
            "flexDirection": "column",
            "fontSize": 12,
          }
        }
      >
         undefined
      </Text>
    </View>
    <View
      style={
        Object {
          "alignSelf": "stretch",
          "flex": 1,
          "flexDirection": "row",
          "justifyContent": "space-between",
          "marginHorizontal": 20,
          "marginTop": 5,
        }
      }
    >
      <Text
        accessible={true}
        allowFontScaling={true}
        disabled={false}
        ellipsizeMode="tail"
        style={
          Object {
            "alignSelf": "center",
            "color": "#CCC",
            "flexDirection": "column",
            "fontSize": 14,
            "fontWeight": "bold",
          }
        }
      >
        Activity Type
      </Text>
      <Text
        accessible={true}
        allowFontScaling={true}
        disabled={false}
        ellipsizeMode="tail"
        style={
          Object {
            "alignSelf": "center",
            "color": "#CCC",
            "flexDirection": "column",
            "fontSize": 12,
          }
        }
      >
        Not Available
      </Text>
    </View>
    <View
      style={
        Object {
          "alignSelf": "stretch",
          "flex": 1,
          "flexDirection": "row",
          "justifyContent": "space-between",
          "marginHorizontal": 20,
          "marginTop": 5,
        }
      }
    >
      <Text
        accessible={true}
        allowFontScaling={true}
        disabled={false}
        ellipsizeMode="tail"
        style={
          Object {
            "alignSelf": "center",
            "color": "#CCC",
            "flexDirection": "column",
            "fontSize": 14,
            "fontWeight": "bold",
          }
        }
      >
        Average Grade
      </Text>
      <Text
        accessible={true}
        allowFontScaling={true}
        disabled={false}
        ellipsizeMode="tail"
        style={
          Object {
            "alignSelf": "center",
            "color": "#CCC",
            "flexDirection": "column",
            "fontSize": 12,
          }
        }
      >
        undefined %
      </Text>
    </View>
    <View
      style={
        Object {
          "alignSelf": "stretch",
          "flex": 1,
          "flexDirection": "row",
          "justifyContent": "space-between",
          "marginHorizontal": 20,
          "marginTop": 5,
        }
      }
    >
      <Text
        accessible={true}
        allowFontScaling={true}
        disabled={false}
        ellipsizeMode="tail"
        style={
          Object {
            "alignSelf": "center",
            "color": "#CCC",
            "flexDirection": "column",
            "fontSize": 14,
            "fontWeight": "bold",
          }
        }
      >
        Climb Category
      </Text>
      <Text
        accessible={true}
        allowFontScaling={true}
        disabled={false}
        ellipsizeMode="tail"
        style={
          Object {
            "alignSelf": "center",
            "color": "#CCC",
            "flexDirection": "column",
            "fontSize": 12,
          }
        }
      >
        Not Available
      </Text>
    </View>
    <View
      style={
        Object {
          "alignSelf": "stretch",
          "flex": 1,
          "flexDirection": "row",
          "justifyContent": "space-between",
          "marginHorizontal": 20,
          "marginTop": 5,
        }
      }
    >
      <Text
        accessible={true}
        allowFontScaling={true}
        disabled={false}
        ellipsizeMode="tail"
        style={
          Object {
            "alignSelf": "center",
            "color": "#CCC",
            "flexDirection": "column",
            "fontSize": 14,
            "fontWeight": "bold",
          }
        }
      >
        Elevation Gain
      </Text>
      <Text
        accessible={true}
        allowFontScaling={true}
        disabled={false}
        ellipsizeMode="tail"
        style={
          Object {
            "alignSelf": "center",
            "color": "#CCC",
            "flexDirection": "column",
            "fontSize": 12,
          }
        }
      >
        undefined meters
      </Text>
    </View>
  </View>
  <View
    style={
      Object {
        "alignSelf": "stretch",
        "flex": 0.1,
        "marginBottom": 10,
      }
    }
  />
</View>
`;
